crates/sui/src/upgrade_compatibility.rs

```diff
@@ -101,7 +109,127 @@ pub(crate) enum UpgradeCompatibilityModeError {
     },
 }
 
+/// A list of errors that can occur during upgrade compatibility checks.
+#[derive(Debug, Clone, Default)]
+pub struct UpgradeErrorList {
+    errors: Vec<UpgradeCompatibilityModeError>,
+    source: Option<String>,
+}
+
+impl UpgradeErrorList {
+    fn push(&mut self, err: UpgradeCompatibilityModeError) {
+        self.errors.push(err);
+    }
+
+    /// Only keep the errors that break compatibility with the given [`Compatibility`]
+    fn retain_incompatible(&mut self, compatibility: &Compatibility) {
+        self.errors
+            .retain(|e| e.breaks_compatibility(compatibility));
+    }
+
+    /// Print the errors to the console with the relevant source code.
+    fn print_errors(
+        &mut self,
+        compiled_unit_with_source: &CompiledUnitWithSource,
+    ) -> Result<(), Error> {
+        for err in self.errors.clone() {
+            match err {
+                UpgradeCompatibilityModeError::StructMissing { name, .. } => {
+                    self.print_missing_definition(
+                        "Struct",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                UpgradeCompatibilityModeError::EnumMissing { name, .. } => {
+                    self.print_missing_definition(
+                        "Enum",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                UpgradeCompatibilityModeError::FunctionMissingPublic { name, .. } => {
+                    self.print_missing_definition(
+                        "Function",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                UpgradeCompatibilityModeError::FunctionMissingEntry { name, .. } => {
+                    self.print_missing_definition(
+                        "Function",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    /// retrieve the source, caches the source after the first read
+    fn source(
+        &mut self,
+        compiled_unit_with_source: &CompiledUnitWithSource,
+    ) -> Result<&String, Error> {
+        if self.source.is_none() {
+            let source_path = compiled_unit_with_source.source_path.clone();
+            let source_content = fs::read_to_string(&source_path)?;
+            self.source = Some(source_content);
+        }
+        Ok(self.source.as_ref().unwrap())
+    }
```

Comment:
The interface here feels strange/error-prone to me: In particular the fact that you need to pass the `compiled_unit_with_source` in to this function in order to get its source, but then the `source` is cached unconditionally (we only check whether we have a source before or not, and not whether that source is for this compiled unit or not).

There are two ways to address this (that I can think of):

- Given that a module belongs to only one source file (today -- although we've talked about easing this in some limited cases) and the compatibility check runs per-module, you could find a way to initialise the error reporter with the source it is checking compatibility for.
- Alternatively, you could have a central structure that caches files, which wraps `SimpleFiles` -- this seems to be what that API nudges us towards, and it means that you maintain this cache across modules (multiple modules can exist in one file).

My preference is towards the second approach.